core/trino-main/src/main/java/io/trino/sql/planner/optimizations/DeterminePartitionCount.java

```diff
@@ -317,32 +319,37 @@ private static double getSourceNodesOutputStats(PlanNode root, ToDoubleFunction<
                 .sum();
     }
 
-    private static boolean isEligibleRemoteExchangePresent(PlanNode root)
+    private static boolean isEligibleRemoteExchangePresent(PlanNode root, boolean taskRetries)
     {
         return PlanNodeSearcher.searchFrom(root)
-                .where(node -> node instanceof ExchangeNode exchangeNode && isEligibleRemoteExchange(exchangeNode))
+                .where(node -> node instanceof ExchangeNode exchangeNode && isEligibleRemoteExchange(exchangeNode, taskRetries))
                 .matches();
     }
 
-    private static boolean isEligibleRemoteExchange(ExchangeNode exchangeNode)
+    private static boolean isEligibleRemoteExchange(ExchangeNode exchangeNode, boolean taskRetries)
     {
         if (exchangeNode.getScope() != REMOTE || exchangeNode.getType() != REPARTITION) {
             return false;
         }
+
         PartitioningHandle partitioningHandle = exchangeNode.getPartitioningScheme().getPartitioning().getHandle();
+
         return !partitioningHandle.isScaleWriters()
                 && !partitioningHandle.isSingleNode()
-                && partitioningHandle.getConnectorHandle() instanceof SystemPartitioningHandle;
+                && partitioningHandle.getConnectorHandle() instanceof SystemPartitioningHandle
+                && (!taskRetries || partitioningHandle == FIXED_HASH_DISTRIBUTION); // for FTE it only makes sense to set partition count fot hash partitioned fragments
```

Comment:
I had hard time to hand craft simple test query which has a HASH distributed stage which produces output consumed by ARBITRARY distributed stage.

Reply:
Oh actually I have - sth one moment